### ISSUE TITLE
1.1.0-beta5 BasedOnUbi check fails with "Error when querying pyxis for uncompressed top layer idsGet"

### ISSUE THREAD

**Issue by tkrishtop** [redhat-openshift-ecosystem/openshift-preflight#508]
## Bug Description
1.1.0-beta5 BasedOnUbi fails with "Error when querying pyxis for uncompressed top layer idsGet"

## Version and Command Invocation
1.1.0-beta5

## Steps to Reproduce:
```
$ PFLT_JUNIT=true PFLT_ARTIFACTS=$PWD/artifacts PFLT_LOGFILE=$PWD//artifacts/preflight.log PFLT_LOGLEVEL=trace ./preflight_1.1.0-beta5 check container quay.io/rh-nfv-int/testpmd-operator@sha256:3e8fc703c71a7ccaca24b7312f8fcb3495370c46e7abc12975757b76430addf5 --certification-project-id=XXX --pyxis-api-token=XXX --docker-config=$PWD/repoconfig/config.json

---snip---

time="2022-03-23T14:14:39+01:00" level=trace msg="the layerIds passed to pyxis are [sha256:476579af086aee78cb12d88b08fd6b61a80a3f055cae293ec2d15a37dbe7b3f9 sha256:1e8cd6732429ccfaea139348807617a5315fd7ca469850423f77f51661e8fe51 sha256:f6d2aa6d27b5382dcd2f3534a017c4e1253a24f099c43a743a3321097c9a5ebf sha256:f2ab5915ba27e95408674610ef2c2ffd23fd224c7014802e5a118327414d7d89 sha256:92b32b08eaa0cc7be98e27dbd763c81cf07725973681dcd0facc06211b9fddbc sha256:4b96680367ad7b7de111ee9c62b33d9eb096d32dad9b8cfe1676c996591efee0 sha256:48c720b7930a9a0b952e7c7843a5fb17f6037fdd0c8c5fdfa9dd65f80b9e9557 sha256:653aa12a30e9661234f0f682ac609bc422444aafcc830a96d48e5ece27079ea9 sha256:b50ba50688e26ce55080cf1b781fc7d3dcacce235a8f6499ebc51805cb11f425 sha256:13dde587dedded645c487505fb6104002bf776f659807ca3e4faf3ed0c520075 sha256:48b366afe5a590835733152cce37f6bd2cd524a7c4af65013b0d2057469854ca sha256:81268765028d22daa394647e571a825a1adfedbbbf42bc0cc403ba5a6efbc939 sha256:b54c8a4f954e68ad5a6c1391bc8a40873e6eaaaaf20f29a16165abf9d7037c30 sha256:30ab82e63506631cc0f71f72255849d33c3270cccfd1549ecf47fa5b9472a7bc sha256:f90d014ac9f6f8bea72df39d4b323daa51d11bacab98a704742903676e1ce38e]"

time="2022-03-23T14:15:39+01:00" level=error msg="Get \"https://catalog.redhat.com/api/containers/v1/images?filter=repositories.registry%3Din%3D%28registry.access.redhat.com%29+and+uncompressed_top_layer_id%3Din%3D%28sha256%3A476579af086aee78cb12d88b08fd6b61a80a3f055cae293ec2d15a37dbe7b3f9%2Csha256%3A1e8cd6732429ccfaea139348807617a5315fd7ca469850423f77f51661e8fe51%2Csha256%3Af6d2aa6d27b5382dcd2f3534a017c4e1253a24f099c43a743a3321097c9a5ebf%2Csha256%3Af2ab5915ba27e95408674610ef2c2ffd23fd224c7014802e5a118327414d7d89%2Csha256%3A92b32b08eaa0cc7be98e27dbd763c81cf07725973681dcd0facc06211b9fddbc%2Csha256%3A4b96680367ad7b7de111ee9c62b33d9eb096d32dad9b8cfe1676c996591efee0%2Csha256%3A48c720b7930a9a0b952e7c7843a5fb17f6037fdd0c8c5fdfa9dd65f80b9e9557%2Csha256%3A653aa12a30e9661234f0f682ac609bc422444aafcc830a96d48e5ece27079ea9%2Csha256%3Ab50ba50688e26ce55080cf1b781fc7d3dcacce235a8f6499ebc51805cb11f425%2Csha256%3A13dde587dedded645c487505fb6104002bf776f659807ca3e4faf3ed0c520075%2Csha256%3A48b366afe5a590835733152cce37f6bd2cd524a7c4af65013b0d2057469854ca%2Csha256%3A81268765028d22daa394647e571a825a1adfedbbbf42bc0cc403ba5a6efbc939%2Csha256%3Ab54c8a4f954e68ad5a6c1391bc8a40873e6eaaaaf20f29a16165abf9d7037c30%2Csha256%3A30ab82e63506631cc0f71f72255849d33c3270cccfd1549ecf47fa5b9472a7bc%2Csha256%3Af90d014ac9f6f8bea72df39d4b323daa51d11bacab98a704742903676e1ce38e%29\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"

time="2022-03-23T14:15:39+01:00" level=error msg="Error when querying pyxis for uncompressed top layer idsGet \"https://catalog.redhat.com/api/containers/v1/images?filter=repositories.registry%3Din%3D%28registry.access.redhat.com%29+and+uncompressed_top_layer_id%3Din%3D%28sha256%3A476579af086aee78cb12d88b08fd6b61a80a3f055cae293ec2d15a37dbe7b3f9%2Csha256%3A1e8cd6732429ccfaea139348807617a5315fd7ca469850423f77f51661e8fe51%2Csha256%3Af6d2aa6d27b5382dcd2f3534a017c4e1253a24f099c43a743a3321097c9a5ebf%2Csha256%3Af2ab5915ba27e95408674610ef2c2ffd23fd224c7014802e5a118327414d7d89%2Csha256%3A92b32b08eaa0cc7be98e27dbd763c81cf07725973681dcd0facc06211b9fddbc%2Csha256%3A4b96680367ad7b7de111ee9c62b33d9eb096d32dad9b8cfe1676c996591efee0%2Csha256%3A48c720b7930a9a0b952e7c7843a5fb17f6037fdd0c8c5fdfa9dd65f80b9e9557%2Csha256%3A653aa12a30e9661234f0f682ac609bc422444aafcc830a96d48e5ece27079ea9%2Csha256%3Ab50ba50688e26ce55080cf1b781fc7d3dcacce235a8f6499ebc51805cb11f425%2Csha256%3A13dde587dedded645c487505fb6104002bf776f659807ca3e4faf3ed0c520075%2Csha256%3A48b366afe5a590835733152cce37f6bd2cd524a7c4af65013b0d2057469854ca%2Csha256%3A81268765028d22daa394647e571a825a1adfedbbbf42bc0cc403ba5a6efbc939%2Csha256%3Ab54c8a4f954e68ad5a6c1391bc8a40873e6eaaaaf20f29a16165abf9d7037c30%2Csha256%3A30ab82e63506631cc0f71f72255849d33c3270cccfd1549ecf47fa5b9472a7bc%2Csha256%3Af90d014ac9f6f8bea72df39d4b323daa51d11bacab98a704742903676e1ce38e%29\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"

time="2022-03-23T14:15:39+01:00" level=error msg="No matching layer ids found in pyxis db. Please verify if the image is based on a recent UBI image"

time="2022-03-23T14:15:39+01:00" level=info msg="check completed: BasedOnUbi" result=FAILED
```

## Expected Result
- BasedOnUbi should not call Pyxis 
- BasedOnUbi should pass, in Preflight 1.0.8 this image was 100% green

```
preflight_container_testpmd-operator_v0.2.9_results.json
{
  "image": "quay.io/rh-nfv-int/testpmd-operator@sha256:3e8fc703c71a7ccaca24b7312f8fcb3495370c46e7abc12975757b76430addf5",
  "passed": true,
  "test_library": {
    "name": "github.com/redhat-openshift-ecosystem/openshift-preflight",
    "version": "1.0.8",
    "commit": "f6eb6893c33c39f5a9ba9e46c0c217f3079dd3ac"
  },
  "results": {
    "passed": [
      {
        "name": "LayerCountAcceptable",
        "description": "Checking if container has less than 40 layers.  Too many layers within the container images can degrade container performance."
      },
      {
        "name": "HasNoProhibitedPackagesMounted",
        "elapsed_time": 186,
        "description": "Checks to ensure that the image in use does not include prohibited packages, such as Red Hat Enterprise Linux (RHEL) kernel packages."
      },
      {
        "name": "HasRequiredLabel",
        "elapsed_time": 108,
        "description": "Checking if the required labels (name, vendor, version, release, summary, description) are present in the container metadata."
      },
      {
        "name": "RunAsNonRoot",
        "description": "Checking if container runs as the root user because a container that does not specify a non-root user will fail the automatic certification, and will be subject to a manual review before the container can be approved for publication"
      },
      {
        "name": "BasedOnUbi",
        "description": "Checking if the container's base image is based upon the Red Hat Universal Base Image (UBI)"
      },
      {
        "name": "HasLicense",
        "description": "Checking if terms and conditions applicable to the software including open source licensing information are present. The license must be at /licenses"
      },
      {
        "name": "HasUniqueTag",
        "elapsed_time": 315,
        "description": "Checking if container has a tag other than 'latest', so that the image can be uniquely identified."
      }
    ],
    "failed": [],
    "errors": []
  }
}
```

## Actual Result
- BasedOnUbi calls Pyxis, and it's not clear why 
- BasedOnUbi fails

**Comment by vikasmulaje** [comment]
@tkrishtop This is an improvement in BasedOnUbi.
Check  https://github.com/redhat-openshift-ecosystem/openshift-preflight/issues/25
https://github.com/redhat-openshift-ecosystem/openshift-preflight/pull/482

**Comment by acornett21** [comment]
@tkrishtop Calling Pyxis for this check is a requirement. I validated the above image passed this check in UAT after prod data was migrated to UAT. There is still an outstanding issues in Pyxis that needs to be pushed to production improve the performance of this call, since `preflight` is calling the external instance of Pyxis. That can be tracked in this [JIRA](https://issues.redhat.com/browse/ISV-1843). I will leave this bug open until that change makes it to production and you can validate.

**Comment by tkrishtop** [comment]
Hi @acornett21, thank you for the response! 

The real issue concerns not this testing image testpmd but the private image of one of our Telco partners. I tried that private image with UAT, there is an empty response from Pyxis, and then everything's fine `msg="check completed: BasedOnUbi" result=PASSED`:
```
time="2022-03-24T17:57:04+01:00" level=trace msg="query response from pyxis {\n  \"data\": [],\n  \"page\": 0,\n  \"page_size\": 100,\n  \"total\": 0\n}\n"
```

Let's wait for the prod refactoring.

**Comment by acornett21** [comment]
The data for the failed image is actually this data:

```
time="2022-03-24T17:57:04+01:00" level=debug msg="running check: BasedOnUbi"
time="2022-03-24T17:57:04+01:00" level=trace msg="the layerIds passed to pyxis are [sha256:476579af086aee78cb12d88b08fd6b61a80a3f055cae293ec2d15a37dbe7b3f9 sha256:1e8cd6732429ccfaea139348807617a5315fd7ca469850423f77f51661e8fe51 sha256:f6d2aa6d27b5382dcd2f3534a017c4e1253a24f099c43a743a3321097c9a5ebf sha256:f2ab5915ba27e95408674610ef2c2ffd23fd224c7014802e5a118327414d7d89 sha256:92b32b08eaa0cc7be98e27dbd763c81cf07725973681dcd0facc06211b9fddbc sha256:4b96680367ad7b7de111ee9c62b33d9eb096d32dad9b8cfe1676c996591efee0 sha256:48c720b7930a9a0b952e7c7843a5fb17f6037fdd0c8c5fdfa9dd65f80b9e9557 sha256:653aa12a30e9661234f0f682ac609bc422444aafcc830a96d48e5ece27079ea9 sha256:b50ba50688e26ce55080cf1b781fc7d3dcacce235a8f6499ebc51805cb11f425 sha256:13dde587dedded645c487505fb6104002bf776f659807ca3e4faf3ed0c520075 sha256:48b366afe5a590835733152cce37f6bd2cd524a7c4af65013b0d2057469854ca sha256:81268765028d22daa394647e571a825a1adfedbbbf42bc0cc403ba5a6efbc939 sha256:b54c8a4f954e68ad5a6c1391bc8a40873e6eaaaaf20f29a16165abf9d7037c30 sha256:30ab82e63506631cc0f71f72255849d33c3270cccfd1549ecf47fa5b9472a7bc sha256:f90d014ac9f6f8bea72df39d4b323daa51d11bacab98a704742903676e1ce38e]"
time="2022-03-24T17:57:04+01:00" level=trace msg="query response from pyxis {\n  \"data\": [],\n  \"page\": 0,\n  \"page_size\": 100,\n  \"total\": 0\n}\n"
time="2022-03-24T17:57:04+01:00" level=error msg="No matching layer ids found in pyxis db. Please verify if the image is based on a recent UBI image"
time="2022-03-24T17:57:04+01:00" level=info msg="check completed: BasedOnUbi" result=FAILED
```

**Comment by tkrishtop** [comment]
Hi @acornett21 I retested this morning:

Pyxis UAT + testpmd + 1.1.0-beta5 = non-zero response from Pyxis = BasedOnUbi passed
Pyxis UAT + testpmd + 1.1.0-beta5 = Pyxis timeout = BasedOnUbi failed --> would it be better to provide more precise error in this situation?

**Comment by tkrishtop** [comment]
@acornett21 I mean the error displayed in the results. Currently, it's confusing for the partners which do base their images on the UBI but receive an error `Change the FROM directive in your Dockerfile or Containerfile to FROM registry.access.redhat.com/ubi8/ubi` because of Pyxis timeout.
```
    "failed": [
      {
        "name": "BasedOnUbi",
        "elapsed_time": 60001,
        "description": "Checking if the container's base image is based upon the Red Hat Universal Base Image (UBI)",
        "help": "Check BasedOnUbi encountered an error. Please review the preflight.log file for more information.",
        "suggestion": "Change the FROM directive in your Dockerfile or Containerfile to FROM registry.access.redhat.com/ubi8/ubi",
        "knowledgebase_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide",
        "check_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide"
      }
```

**Comment by bcrochet** [comment]
That part is fixed by #518. Which will be in the next beta. It will now be reported as an error, not a failure.

**Comment by bcrochet** [comment]
With the release of 1.1.0-beta6, I'm marking this as closed. Feel free to reopen if necessary.
